README.md
# transRedBatch
Batch Translation assist for Translator++.

This tool is based on the general translation guide, [which can be found here](TranslationGuide.MD). There is also a very short Cheat Sheet that is a condensed reference to the guide, [the cheatsheet can be found here](CheatSheet.MD).

## What does the Addon do?

The addon was initially created as a replacement for T++'s default Batch Translation. The default Batch translator is somewhat slow due to how the text is added back into the project.

This addon solves that speed issue (which is notable on bigger projects) by caching the position of every line being translated - so there is no search to put them back in.

![The batch translation window](/guide/batchTranslator.png)

The addon also has some helper functions to help in the process of batch machine translating projects. The buttons are added to T++ in a red color, and can be found at the top:

![The addon's buttons](/guide/batchButtons.png)

## Every function

+ ![Prepare](/guide/button1Prepare.png) - The first button is the "Prepare Project for Batch Translation". This button only works for RPG Maker projects, what it does is follow the guide above and apply tags.
    + Safe to translate contexts will be kept without any marks. Contexts which are known to be translatable, but might cause problems will be tagged Yellow. Contexts which are unlikely to be translatable will be tagged Red. Additionally, contexts which are known to not be visible will be removed from the project entirely (these are event names, animation names, etc, text that exists but is never seen in-game).
+ ![Translate](/guide/button2Translate.png) - The second button opens the batch translation window above.
+ ![Wrap](/guide/button3Wrap.png) - The third button will word wrap the text inside the project following configurable rules from the addon's option window.
    + This word wrapping is very different to the default word wrapping. Instead of measuring character count, we measure character width. This means that the length settings should be set in full-width characters. For instance, RPG Maker MV games default to showing up to 28 japanese full-width characters per line. So that's the default setting this addon has.

Examples for text width:

+ 平仮名ひらがな片仮名カタカナ
    + 14 characters
+ 0123456789012345678901234
    + 24 characters (but occupy slightly less space)
+ Abcde fghi. Jklmn opqrs?
    + 24 characters, but even shorter.

Calculating how much width each sentence has is far too expensive (computationally), so instead we approximate that by counting how much width each character has (which is close enough of an approximation). This allows us to wrap sentences much closer to the edges of message boxes/etc.

The way the wrapping is performed can also be found in T++'s options panel.

Changelog.MD
# Version 0.16
+ Prevent infinite loops by sacricifing padding in case it is too long
+ Add configurable options
+ Add option to try and reduce line amount in critical contexts / e.g. descriptions
    + This is because descriptions can't span into the next window if they don't fit, so we only have 4 or so lines depending on the game. If this saves us a single line that can mean the difference between the text rendering out of bounds or in.
+ God protect us from Comments
+ Fix: Breaking descriptions now works
+ New Option: Allow merging sentences regardless of it being part of the contexts.
    + Because this merges sentence tails with next sentences, it might result in nicer looking text in general.
+ Faster searching through contexts
+ Merging sentences now works as desired - can target specific contexts and they will have the size that they should have normally, too.

# Version 0.13
+ Recovers padding after wrapping text.
+ No longer using default Word Wrap implementation.
+ New wrap implementation counts character width rather than character count.
    + It defaults to MV character size at default font. Might need to tinker for other games/engine.

# Version 0.12
+ Fix: Correct default translator
+ Fix: It correctly loads now.


# Version 0.10
+ Fix: Translators sorted badly.
+ Fix: Grid not updated after wrapping
+ Fix: Grid not updated after aborting translations
+ Some more VX Ace on old parser support
+ Try to order the buttons
+ Removed most unnecessary performance checks.
+ Known bugs: Wrong translator selected by default.

# Version 0.09
+ Remove minimum height from the batch window.

# Version 0.08
+ Gets an actual option window when using.
    + While it supports any translators, some of the non-red translators might behave erratically when called this way. It's mostly there as a "well, there it is". If not using red translator engines I'd recommend just using the default batch translator for full support.

# Version 0.07
+ Saving the project is now done asynchronously. This is a massive speed gain on small projects, but a pleasant speed gain on projects of any size.
    + This also means that there is less time wasted saving the project between batches, so uptime is higher when the bottleneck is the translation process itself.

# Version 0.06
+ Switches the awful links for toolbar buttons

# Version 0.05
+ Better word wrap defaults, might still be too big if too many big letters are used. Only japanese gets monospace, I'm afraid.
+ Now marks Game.ini as Red in VX Ace games.

# Version 0.04
+ Added basic VX Ace support for the Prepare Project. Assumes New Parser for now.
    + A little support for the old parser but no promises!
+ Added button for Word Wrap. Should work relatively well on RPG Maker games of all kinds. No promises!
    + For MV/MZ, it's still better to just leave them as-is and use a word-wrap plugin instead... if the game supports it!

# Version 0.03
+ Capable of "Preparing Project", which works through rows following the guide automatically.
    + Prepare Project, for now, only works on MV/MZ games. Games originally made with English Characters require an additional manual look at Plugins.js, if translating Scripts.
    + It is recommended to use Custom Escapers and Cut Corners if translating scripts, as script text position isn't allowed to change!
+ Added placeholder buttons
+ TODO: Add better buttons.

# Version 0.02
+ Added Changelog
+ Added Guide
+ Added Cheat Sheet
+ TODO: Add cheat sheet inside the addon.

# Version 0.01
+ "It works" the release

ts/Main.ts
/// <reference path="RedBatchTranslator.ts" />
/// <reference path="RedButtonManager.ts" />

var thisAddon = <any> this;

enum thisAddonOptions {
    alwaysMerge = "alwaysMerge",
    fontFamily = "fontFamily",
    longContexts = "longContexts",
    shortContexts = "shortContexts",
    mergeContexts = "mergeContexts",
    longLength = "longLength",
    shortLength = "shortLength"
}

thisAddon.optionsForm = {
    [thisAddonOptions.fontFamily] : {
        "type": "string",
        "title": "Font Family",
        "description": "When wrapping texts, this is the font which will be used to decide length. Ideally, use one that looks as similar as possible to the game's.",
        "default" : "Arial",
        "HOOK": "thisAddon.config." + thisAddonOptions.fontFamily
      },
    [thisAddonOptions.mergeContexts] : {
        "type": "string",
        "title": "Merge Wrap Contexts",
        "description": "Instead of breaking lines into smaller lines, the remainder of one line will be added to the start of the next line (after padding) in these contexts. This should result in less overall lines, which is good when you don't have the benefit of just making a second message.",
        "HOOK": "thisAddon.config." + thisAddonOptions.mergeContexts,
        "default" : [
            "description"
        ].join(",")
    },
    [thisAddonOptions.alwaysMerge] : {
        "type": "boolean",
        "title": "Always Merge",
        "description": `Applies the processing of merging with the next line on every wrappable context, for those who like compact text.`,
        "default" : false,
        "HOOK": "thisAddon.config." + thisAddonOptions.alwaysMerge
      },
    [thisAddonOptions.shortContexts] : {
        "type": "string",
        "title": "Short Contexts",
        "description": "These contexts will wrap to the short length type. These should be used when showing pictures, or when the game simply has a smaller message box.",
        "default" : [
            "hasPicture", //"Dialogue", // Why a uppercase DIalogue? I don't remember! Need to find out if this is correct.
        ].join(","),
        "HOOK": "thisAddon.config." + thisAddonOptions.shortContexts
      },
      [thisAddonOptions.shortLength] : {
          "type": "number",
          "title": "Short Length",
          "description": `The length to wrap short contexts to. This length is measured in square, monospace characters, using japanese characters as basis.\nWhat this means is that you are defining how long your sentences can be in amount of japanese characters. This is useful because proportional fonts will change width depending on the characters used, so wrapping to character count isn't very accurate. 24 is the number of japanese characters that fit in default MV message window with picture.\nCounting assist: ０１２３４５６７８９`,
          "default" : 24,
          "HOOK": "thisAddon.config." + thisAddonOptions.shortLength
        },
    [thisAddonOptions.longContexts] : {
        "type": "string",
        "title": "Long Wrap Contexts",
        "description": "These contexts will wrap to the long length. So ideally these are the full screen contexts.",
        "HOOK": "thisAddon.config." + thisAddonOptions.longContexts,
        "default" : [
            "dialogue", "message1", "message2", "message3", "description",
            "message", "noPicture", "scrollingMessage"
        ].join(",")
    },
    [thisAddonOptions.longLength] : {
        "type": "number",
        "title": "Long Length",
        "description": `The length to wrap long contexts to. This length is measured in square, monospace characters, using japanese characters as basis.\nWhat this means is that you are defining how long your sentences can be in amount of japanese characters. This is useful because proportional fonts will change width depending on the characters used, so wrapping to character count isn't very accurate. 28 is the number of japanese characters that fit in default MV message window without picture.\nCounting assist: ０１２３４５６７８９`,
        "default" : 28,
        "HOOK": "thisAddon.config." + thisAddonOptions.longLength
      },
};

thisAddon.config = thisAddon.config == undefined ? {} : thisAddon.config;;

for (let id in thisAddon.optionsForm) {
    if (typeof thisAddon.config[id] != typeof thisAddon.optionsForm[id].default) {
        if (typeof thisAddon.config[id] == "string" && typeof thisAddon.optionsForm[id].default == "number") {
            thisAddon.config[id] = parseInt(thisAddon.config[id]);
        } else {
            thisAddon.setConfig(id, thisAddon.optionsForm[id].default);
        }
    }
}

function cleanContexts (str : string) {
    str = str.replaceAll(/[,\.;\-]/g, ",").replaceAll(/ /g, "").toLowerCase();
    return str.split(",");
}

const measurer : HTMLSpanElement = document.createElement("pre");
measurer.style.fontSize = "24px";
measurer.style.fontFamily = "Arial";
measurer.style.setProperty("all", "revert", "important");
measurer.style.setProperty("position", "position", "important");
measurer.style.setProperty("visibility", "hidden", "important");
measurer.style.setProperty("width", "max-content", "important");

function measureTextPxWidth(text : string) {
    measurer.innerText = text;
    document.body.appendChild(measurer);
    let width = measurer.getBoundingClientRect().width;
    document.body.removeChild(measurer);
    console.log(`Width of text "${text}" is ${width}`);
    return width;
}



let japaneseLength = measureTextPxWidth("。");
let wordWrapNoPicture = 28 * japaneseLength; // MV Size is 28 characters, have to check against the others. 
let wordWrapPicture = 24 * japaneseLength;

function updateValues () {
    if (measurer.style.fontFamily != thisAddon.config[thisAddonOptions.fontFamily]) {
        measurer.style.fontFamily = thisAddon.config[thisAddonOptions.fontFamily];
        sizes = {};
    }
    japaneseLength = measureTextPxWidth("。");
    wordWrapNoPicture = parseInt(thisAddon.config[thisAddonOptions.longLength]) * japaneseLength;
    wordWrapNoPicture = parseInt(thisAddon.config[thisAddonOptions.shortLength]) * japaneseLength;
}


function getPadAtLine (line : string) : string {
    if (line === undefined) {
        return "";
    }
    let match = line.match(/^\s+/);
    if (match !== null) {
        return match[0];
    } else {
        return "";
    }
}

/* function equatePad (origin : string, translation : string) {
    let splitTranslation = translation.split(/\r?\n/g);
    let splitOrigin = origin.split(/\r?\n/g);

    for (let line = 0; line < splitTranslation.length; line++) {
        splitTranslation[line] = getPadAtLine(splitOrigin, line) + splitTranslation[line].trim();
    }

    return splitTranslation.join("\n");
} */

let sizes : {[id : string] : number} = {};

function getSize (text : string) {
    if (sizes[text] == undefined) {
        sizes[text] = measureTextPxWidth(text);
    }
    return sizes[text];
}

function isTrue (value : any) {
    return (value === "1" || value === "true" || value === 1 || value === true);
}

function wrapTo (text : string, length: number, mergeWithNext : boolean) {
    let lines = text.split(/\r?\n/g);
    let newLines = [];
    for (let i = 0; i < lines.length; i++) {
        let currentSize = 0;
        let originalPad = getPadAtLine(lines[i]);
        let nextPad = getPadAtLine(lines[i + 1]);
        if (nextPad.length < originalPad.length) {
            nextPad = originalPad;
        }
        let padSize = getSize(nextPad);
        if (padSize >= length/2) {
            // the pad is too large for us to work with
            nextPad = "  ";
        }
        let charAt : number;
        for (charAt = 0; charAt < lines[i].length; charAt++) {
            currentSize += getSize(lines[i].charAt(charAt));
            if (currentSize > length) {
                break;
            }
        }
        
        if (currentSize > length) {
            let closestSpace = lines[i].lastIndexOf(" ", charAt);
            if (closestSpace <= originalPad.length) {
                // Just one bigassword? Fine, suit yourself.
                closestSpace = charAt;
            }
            newLines.push(lines[i].substring(0, closestSpace));
            if (charAt < lines[i].length) {
                if (mergeWithNext || isTrue(thisAddon.config[thisAddonOptions.alwaysMerge])) {
                    if (lines[i + 1] === undefined) {
                        lines[i + 1] = "";
                    }
                    lines[i + 1] = nextPad + lines[i].substring(closestSpace).trim() + " " + lines[i + 1].trim();
                } else {
                    lines[i] = nextPad + lines[i--].substring(closestSpace).trim();
                }
            }
        } else {
            newLines.push(lines[i]);
        }
    }
    return newLines.join("\n");
}

$(document).ready(() => {
    trans.RedBatchTranslatorInstance = new RedBatchTranslator();

    let buttonContainer = document.body.getElementsByClassName("toolbar-content toolbar10 redToolbar")[0];
    if (buttonContainer == undefined) {
        let toolbarContainer = document.body.getElementsByClassName("toolbar mainToolbar")[0];
        buttonContainer = document.createElement("div");
        buttonContainer.className = "toolbar-content toolbar10 redToolbar";
        toolbarContainer.appendChild(buttonContainer);
    }

    let prepareButton = new RedButtonManagerButton("prepareProject", "icon-tasks", "Prepare Project for Batch Translation", () => {
        trans.batchCheckSheet.checkProject();
	});

	let translateButton = new RedButtonManagerButton("batchTranslate", "icon-language-1", "Batch Translate", () => {
        trans.RedBatchTranslatorInstance.open();
	});

    let wrapButton = new RedButtonManagerButton("wrapProject", "icon-commenting", "Wrap Project", () => {
        updateValues();
        let mergeable = new RegExp(`(${thisAddon.config[thisAddonOptions.mergeContexts].split(",").join(")|(")})`);
        let wrappableLong = new RegExp(`(${thisAddon.config[thisAddonOptions.longContexts].split(",").join(")|(")})`);
        let wrappableShort = new RegExp(`(${thisAddon.config[thisAddonOptions.shortContexts].split(",").join(")|(")})`);

        let files = <string[]> trans.getAllFiles();
        for (let i = 0; i < files.length; i++) {
            let file = files[i];
            console.log("Wrapping " + file + " - Progress: " + (i + 1) + "/" + files.length);
            let data = trans.project.files[file].data;
            let contexts = trans.project.files[file].context;
            for (let k = 0; k < data.length; k++) {
                let row = data[k];
                let context = contexts[k];
                // Skip empty / untranslated lines
                if (context == undefined || row == undefined || context.length == 0 || row[1] == undefined || row[1] == "") continue;
                let contextString : string = context.join().toLowerCase();
                let shouldMerge = contextString.match(mergeable) != null;
                if (contextString.match(wrappableLong) != null) {
                    trans.project.files[file].data[k][2] = wrapTo(row[1], wordWrapNoPicture, shouldMerge);
                } else if (contextString.match(wrappableShort) != null) {
                    trans.project.files[file].data[k][2] = wrapTo(row[1], wordWrapPicture, shouldMerge);
                }
            }
        }
        trans.refreshGrid();
        trans.evalTranslationProgress();
	});

    let $buttonContainer = $(buttonContainer);
	$buttonContainer.prepend(wrapButton.getButton());
	$buttonContainer.prepend(translateButton.getButton());
	$buttonContainer.prepend(prepareButton.getButton());
});



/* //@ts-expect-error
window.runTest = (testCount : number, minLength : number, maxLength : number) => {
    //let font = "　０ １ ２ ３ ４ ５ ６ ７ ８ ９ Ａ Ｂ Ｃ Ｄ Ｅ Ｆ Ｇ Ｈ Ｉ Ｊ Ｋ Ｌ Ｍ Ｎ Ｏ Ｐ Ｑ Ｒ Ｓ Ｔ Ｕ Ｖ Ｗ Ｘ Ｙ Ｚ ａ ｂ ｃ ｄ ｅ ｆ ｇ ｈ ｉ ｊ ｋ ｌ ｍ ｎ ｏ ｐ ｑ ｒ ｓ ｔ ｕ ｖ ｗ ｘ ｙ ｚ ， ． ： ； ！ ？ ＂ ＇ ｀ ＾ ～ ￣ ＿ ＆ ＠ ＃ ％ ＋ － ＊ ＝ ＜ ＞ （ ） ［ ］ ｛ ｝ ｟ ｠ ｜ ￤ ／ ＼ ￢ ＄ ￡ ￠ ￦ ￥";
    let font = " 0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

    let texts = [];

    let results = [];
    let differences = [];
    for (let i = 0; i < testCount; i++) {
        let length = Math.random() * maxLength;
        if (length < minLength) {
            length = minLength;
        }
        let text = "";
        for (let k = 0; k < length; k++) {
            text += font[Math.floor(Math.random() * font.length)];
        }

        let realSize = getSize(text);

        let fastSize = 0;
        for (let k = 0; k < text.length; k++) {
            fastSize += getSize(text[k]);
        }

        let diff = Math.abs(fastSize - realSize);
        let diffPerc = Math.round((1000 * diff) / realSize)/10;
        differences.push(diffPerc);

        results.push(`"${text}" - Real Size: ${realSize} - Fast Cache Size: ${fastSize} - Difference: ${diffPerc}%`);
    }

    return [results, differences];
} */
// Test results: 0.1% difference for basic text
// Around 0.5% difference for complex text involving mix of full-width and proportional characters
// Verdict: it's good.